File: leetcode/74-search-a-2d-matrix.py
# https://leetcode.com/problems/search-a-2d-matrix/

from typing import List

class Solution:
	def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
		first = 0 # first row (row with index 0)
		last = len(matrix)-1 # last row (row with index len(matrix)-1)
		
		# binary search for the row that maybe contain the target value
		while first <= last:
			mid_row = first + (last-first) // 2
			if matrix[mid_row][0] <= target and target <= matrix[mid_row][-1]:
				# if target is exist then it should be in this row
				# so we do another binary search but this time inside the row itself
				left = 0
				right = len(matrix[0]) - 1
				while left <= right:
					mid_col = left + (right - left) // 2
					if target == matrix[mid_row][mid_col]:
						return True
					elif target > matrix[mid_row][mid_col]:
						left = mid_col + 1
					else:
						right = mid_col - 1
				break
			elif target < matrix[mid_row][0]:
				last = mid_row - 1
			elif target > matrix[mid_row][-1]:
				first = mid_row + 1

		return False

File: leetcode/74-search-a-2d-matrix.js
// https://leetcode.com/problems/search-a-2d-matrix/

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {

  let m = matrix.length;
  let n = matrix[0].length;

  let top = 0; // top row pointer
  let bottom = m - 1; // bottom row pointer

  while (top <= bottom) {
    // try to identify the row that should contain the target (if it exist)

    const vMid = top + Math.floor((bottom - top) / 2); // vertical middle

    if (matrix[vMid][0] <= target && target <= matrix[vMid][n-1]) {
      // if target value exist it should be in this row ( matrix[vMid] )
      let left = 0;
      let right = n - 1;

      while (left <= right) {
        const hMid = left + Math.floor((right - left) / 2); // horizantal middle
        if (target === matrix[vMid][hMid]) {
          return true;
        } else if (target > matrix[vMid][hMid]) {
          left = hMid + 1;
        } else {
          right = hMid - 1;
        }
      }

      // at this point the target does not exist we should break the outer loop 
      // ( if we don't we will stack in an infinite loop )
      break;
      
    } else if ( matrix[vMid][n-1] < target ) {
      top = vMid + 1;
    } else {
      bottom = vMid - 1;
    }
  }

  return false;

};
